### PR TITLE
Add keys function to get all keys matching pattern

### PR DESCRIPTION
**Pull request by U-Prince**

### CODE REVIEW

File: lib/memclient.js [hunk 1]
```diff
@@ -36,6 +36,18 @@ const cacheProto = {
     callback(null, this._cache.has(mapKey(key, this._segment)))
   },
 
+  keys: function (pattern, callback) {
+    const _key = mapKey(pattern, this._segment)
+    let it = this._cache.keys(_key)
```
Comment (jsumners) [inline lib/memclient.js:41]: Shouldn't this be `const`?

File: lib/memclient.js [hunk 2]
```diff
@@ -36,6 +36,18 @@ const cacheProto = {
...
+    let res = it.next()
+    let result = []
```
Comment (jsumners) [inline lib/memclient.js:43]: Shouldn't this be `const`?

**Review by jsumners — CHANGES_REQUESTED** [review]
Please explain why this function should be added to the generic interface module `abstract-cache`. The purpose of the `abstract-cache` module is to define a baseline protocol that other modules can implement and expand upon. It seems to me you want this feature in `abstract-cache-redis`. If that is the case, you should update your PR there with this level of thoroughness.

Also, this is missing documentation.

File: lib/memclient.js (hunk 2 above)
Comment (U-Prince) [reply lib/memclient.js:43]: Yes, it should be. I'll update it

**Comment by jsumners** [comment]
Adding new methods to this module has implications for _all_ modules that implement the protocol defined by this module. Unless I don't remember what I wrote correctly (highly likely!) then I do not think you have to specifically use `abstract-cache` with `fastify-caching`. As long as the cache provider implements the `abstract-cache` protocol then you can supply it to `fastify-caching`. Which means you can instantiate `abstract-cache-redis` with an already connected client and pass that instance to `fastify-caching`. Thus, this PR can live solely on the `abstract-cache-redis` module. Do you concur?

**Comment by jsumners** [comment]
You do not have to supply an instance of `abstract-cache` to `fastify-caching`. All it requires is that the cache object implement the `abstract-cache` protocol. So you can directly submit an instance of `abstract-cache-redis` to `fastify-caching`. Thus, this PR should be on `abstract-cache-redis`.

Any methods implemented on this module have broader implications for _all_ modules that implement the protocol this module defines.

**Comment by jsumners** [comment]
You do not have to supply an instance of `abstract-cache` to `fastify-caching`. All it requires is that the cache object implement the `abstract-cache` protocol. So you can directly submit an instance of `abstract-cache-redis` to `fastify-caching`. Thus, this PR should be on `abstract-cache-redis`.

Any methods implemented on this module have broader implications for _all_ modules that implement the protocol this module defines.

**Comment by U-Prince** [comment]
> Please explain why this function should be added to the generic interface module `abstract-cache`. The purpose of the `abstract-cache` module is to define a baseline protocol that other modules can implement and expand upon. It seems to me you want this feature in `abstract-cache-redis`. If that is the case, you should update your PR there with this level of thoroughness.
> 
> Also, this is missing documentation.

Any additional API used in abstract-cache-redis, which is not mandated in abstract-cache protocol are only useful if you create the client with connection configuration instead of an already connected Redis client. 
But we are using fastify-caching, and want to make use of the already connected redis client, I have added the keys functionality in both the libraries. Adding the feature only to abstract-cache-redis does not serve the purpose.

**Comment by U-Prince** [comment]
> Adding new methods to this module has implications for _all_ modules that implement the protocol defined by this module. Unless I don't remember what I wrote correctly (highly likely!) then I do not think you have to specifically use `abstract-cache` with `fastify-caching`. As long as the cache provider implements the `abstract-cache` protocol then you can supply it to `fastify-caching`. Which means you can instantiate `abstract-cache-redis` with an already connected client and pass that instance to `fastify-caching`. Thus, this PR can live solely on the `abstract-cache-redis` module. Do you concur?

Yes, I agree about the passing of abstract-cache-redis as an instance to fastify-caching. I am sorry for the confusion. 
I have raised a pull request on the abstract-cache-redis with the desired modifications as well.
But how would adding a new method have implications on the modules implementing this protocol.

**Comment by jsumners** [comment]
> But how would adding a new method have implications on the modules implementing this protocol.

This module, `abstract-cache`, defines the _minimum_ feature set every compliant module must implement. Please see the documentation -- https://github.com/jsumners/abstract-cache/tree/7f712d4e7573db6022379e85ce85f944e2720e6f#protocol . If this`.keys` method were added, then it would need to be added to that list. Thus, this PR would be a semver major change and any implementing modules would need to add support for `.keys` (which may not be possible since that is very much a Redis concept).

**Comment by jsumners** [comment]
I think we can close this given the recent release of `abstract-cache-redis`.

**Pull request closed without merging**